dbms/src/Operators/AggregateContext.cpp

```diff
@@ -31,21 +32,52 @@ void AggregateContext::initBuild(const Aggregator::Params & params, size_t max_t
     }
 
     aggregator = std::make_unique<Aggregator>(params, log->identifier());
-    aggregator->setCancellationHook(std::move(hook));
+    aggregator->setCancellationHook(is_cancelled);
     aggregator->initThresholdByAggregatedDataVariantsSize(many_data.size());
-    inited_build = true;
+    status = AggStatus::build;
     build_watch.emplace();
     LOG_TRACE(log, "Aggregate Context inited");
 }
 
 void AggregateContext::buildOnBlock(size_t task_index, const Block & block)
 {
-    RUNTIME_CHECK(inited_build && !inited_convergent);
+    assert(status.load() == AggStatus::build);
     aggregator->executeOnBlock(block, *many_data[task_index], threads_data[task_index]->key_columns, threads_data[task_index]->aggregate_columns);
     threads_data[task_index]->src_bytes += block.bytes();
     threads_data[task_index]->src_rows += block.rows();
 }
 
+bool AggregateContext::hasSpilledData() const
+{
+    assert(status.load() == AggStatus::build);
+    return aggregator->hasSpilledData();
+}
+
+bool AggregateContext::needSpill(size_t task_index, bool try_mark_need_spill)
+{
+    assert(status.load() == AggStatus::build);
+    auto & data = *many_data[task_index];
+    if (try_mark_need_spill && !data.need_spill)
+        data.tryMarkNeedSpill();
+    return data.need_spill;
+}
+
+void AggregateContext::spillData(size_t task_index)
+{
+    assert(status.load() == AggStatus::build);
+    aggregator->spill(*many_data[task_index]);
+}
+
+LocalAggregateRestorerPtr AggregateContext::buildLocalRestorer()
+{
+    assert(status.load() == AggStatus::build);
+    aggregator->finishSpill();
+    LOG_INFO(log, "Begin restore data from disk for local aggregation.");
+    auto input_streams = aggregator->restoreSpilledData();
```

Comment:
If input_stream size == 0, we can skip creating LocalAggregateRestorer, add the operator will not switch to io pool

Reply:
good point!
Since `aggregator.hasSpilledData` is true, `input_streams` must not be empty.
`RUNTIME_CHECK(!input_streams.empty)` added.